File: Advanced/secretsanta-cordapp/workflows/src/test/java/net/corda/samples/secretsanta/CreateSantaSessionFlowTests.java
package net.corda.samples.secretsanta;

import com.google.common.collect.ImmutableList;
import net.corda.core.identity.CordaX500Name;
import net.corda.samples.secretsanta.contracts.SantaSessionContract;
import net.corda.samples.secretsanta.flows.CreateSantaSessionFlow;
import net.corda.samples.secretsanta.states.SantaSessionState;
import net.corda.core.concurrent.CordaFuture;
import net.corda.core.contracts.Command;
import net.corda.core.contracts.TransactionState;
import net.corda.core.node.NetworkParameters;
import net.corda.core.transactions.SignedTransaction;
import net.corda.testing.node.*;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.time.Instant;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedHashMap;

import static org.jgroups.util.Util.assertEquals;
import static org.jgroups.util.Util.assertTrue;
import static org.junit.jupiter.api.Assertions.assertNull;

public class CreateSantaSessionFlowTests {
    private MockNetwork network;

    private StartedMockNode santa;
    private StartedMockNode elf;

    private NetworkParameters testNetworkParameters = new NetworkParameters(4, Arrays.asList(), 10485760, (10485760 * 5), Instant.now(),1, new LinkedHashMap<>());

    private ArrayList<String> playerNames = new ArrayList<>(Arrays.asList("david", "alice", "bob", "charlie", "olivia", "peter"));
    private ArrayList<String> playerEmails = new ArrayList<>(Arrays.asList("david@corda.net", "alice@corda.net", "bob@corda.net", "charlie@corda.net", "olivia@corda.net", "peter@corda.net"));

    @Before
    public void setup() {

        network = new MockNetwork(new MockNetworkParameters().withCordappsForAllNodes(ImmutableList.of(
                TestCordapp.findCordapp("net.corda.samples.secretsanta.contracts"),
                TestCordapp.findCordapp("net.corda.samples.secretsanta.flows"))).withNetworkParameters(testNetworkParameters)
                .withNotarySpecs(ImmutableList.of(new MockNetworkNotarySpec(CordaX500Name.parse("O=Notary,L=London,C=GB"))))
        );

        santa = network.createPartyNode(null);
        elf = network.createPartyNode(null);

        network.runNetwork();
    }

    @After
    public void tearDown() {
        network.stopNodes();
    }

    @Test
    public void flowUsesCorrectNotary() throws Exception {
        CreateSantaSessionFlow f1 = new CreateSantaSessionFlow(playerNames, playerEmails, elf.getInfo().getLegalIdentities().get(0));
        CordaFuture<SignedTransaction> future = santa.startFlow(f1);
        network.runNetwork();
        SignedTransaction signedTransaction = future.get();
        assertEquals(1, signedTransaction.getTx().getOutputStates().size());

        // ensure correct notary is used
        assertEquals(network.getNotaryNodes().get(0).getInfo().getLegalIdentities().get(0), signedTransaction.getNotary());
    }

    @Test
    public void canCreateSession() throws Exception {
        CreateSantaSessionFlow f1 = new CreateSantaSessionFlow(playerNames, playerEmails, elf.getInfo().getLegalIdentities().get(0));
        CordaFuture<SignedTransaction> future = santa.startFlow(f1);
        network.runNetwork();
        SignedTransaction signedTransaction = future.get();
        assertEquals(1, signedTransaction.getTx().getOutputStates().size());

        SantaSessionState output = signedTransaction.getTx().outputsOfType(SantaSessionState.class).get(0);
        // get some random data from the output to verify
        assertEquals(playerNames, output.getPlayerNames());
    }

    @Test
    public void transactionConstructedHasCorrectOutput() throws Exception {
        CreateSantaSessionFlow f1 = new CreateSantaSessionFlow(playerNames, playerEmails, elf.getInfo().getLegalIdentities().get(0));
        CordaFuture<SignedTransaction> future = santa.startFlow(f1);
        network.runNetwork();
        SignedTransaction signedTransaction = future.get();

        assertEquals(1, signedTransaction.getTx().getOutputStates().size());
        TransactionState tOutput = signedTransaction.getTx().getOutputs().get(0);

        // ensure correct notary is used is used
        assertEquals(network.getNotaryNodes().get(0).getInfo().getLegalIdentities().get(0), tOutput.getNotary());

        SantaSessionState output = signedTransaction.getTx().outputsOfType(SantaSessionState.class).get(0);

        // checking player names, emails, and assignments.
        assertEquals(playerNames, output.getPlayerNames());
        assertEquals(playerEmails, output.getPlayerEmails());
        assertEquals(playerEmails.size(), output.getAssignments().size());
    }


    @Test
    public void transactionConstructedHasOneOutputUsingTheCorrectContract() throws Exception {
        CreateSantaSessionFlow f1 = new CreateSantaSessionFlow(playerNames, playerEmails, elf.getInfo().getLegalIdentities().get(0));
        CordaFuture<SignedTransaction> future = santa.startFlow(f1);
        network.runNetwork();

        SignedTransaction signedTransaction = future.get();

        assertEquals(1, signedTransaction.getTx().getOutputStates().size());
        TransactionState output = signedTransaction.getTx().getOutputs().get(0);

        assertEquals("net.corda.samples.secretsanta.contracts.SantaSessionContract", output.getContract());
    }


    @Test
    public void transactionConstructedByFlowHasOneIssueCommand() throws Exception {
        CreateSantaSessionFlow f1 = new CreateSantaSessionFlow(playerNames, playerEmails, elf.getInfo().getLegalIdentities().get(0));
        CordaFuture<SignedTransaction> future = santa.startFlow(f1);
        network.runNetwork();

        SignedTransaction signedTransaction = future.get();

        assertEquals(1, signedTransaction.getTx().getCommands().size());
        Command command = signedTransaction.getTx().getCommands().get(0);

        assert (command.getValue() instanceof SantaSessionContract.Commands.Issue);
    }

    @Test
    public void transactionConstructedByFlowHasOneCommandWithTheIssuerAndTheOwnerAsASigners() throws Exception {
        CreateSantaSessionFlow f1 = new CreateSantaSessionFlow(playerNames, playerEmails, elf.getInfo().getLegalIdentities().get(0));
        CordaFuture<SignedTransaction> future = santa.startFlow(f1);
        network.runNetwork();

        SignedTransaction signedTransaction = future.get();

        assertEquals(1, signedTransaction.getTx().getCommands().size());
        Command command = signedTransaction.getTx().getCommands().get(0);

        assertEquals(2, command.getSigners().size());
        assertTrue(command.getSigners().contains(santa.getInfo().getLegalIdentities().get(0).getOwningKey()));
        assertTrue(command.getSigners().contains(elf.getInfo().getLegalIdentities().get(0).getOwningKey()));
    }

    @Test
    public void transactionConstructedByFlowHasNoInputsAttachmentsOrTimeWindows() throws Exception {
        CreateSantaSessionFlow f1 = new CreateSantaSessionFlow(playerNames, playerEmails, elf.getInfo().getLegalIdentities().get(0));
        CordaFuture<SignedTransaction> future = santa.startFlow(f1);
        network.runNetwork();

        SignedTransaction signedTransaction = future.get();
        assertEquals(0, signedTransaction.getTx().getInputs().size());
        assertEquals(1, signedTransaction.getTx().getOutputs().size());

        // The single attachment is the contract attachment.
        assertEquals(1, signedTransaction.getTx().getAttachments().size());
        assertNull(signedTransaction.getTx().getTimeWindow());
    }

}

File: kotlin-samples/Advanced/secretsanta-cordapp/workflows/src/test/kotlin/net/corda/samples/secretsanta/CreateSantaSessionFlowTests.kt
package net.corda.samples.secretsanta


import net.corda.core.crypto.SecureHash
import net.corda.core.identity.CordaX500Name
import net.corda.core.node.NetworkParameters
import net.corda.core.transactions.SignedTransaction
import net.corda.samples.secretsanta.contracts.SantaSessionContract
import net.corda.samples.secretsanta.flows.CreateSantaSessionFlow
import net.corda.samples.secretsanta.states.SantaSessionState
import net.corda.testing.node.*
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.jupiter.api.Assertions
import java.time.Instant
import java.util.*


class CreateSantaSessionFlowTests {
    lateinit var network: MockNetwork
    lateinit var santa: StartedMockNode
    lateinit var elf: StartedMockNode

    private val testNetworkParameters = NetworkParameters(4, Arrays.asList(), 10485760, 10485760 * 5, Instant.now(), 1, LinkedHashMap<String, List<SecureHash>>())
    private val playerNames = ArrayList(Arrays.asList("david", "alice", "bob", "charlie", "olivia", "peter"))
    private val playerEmails = ArrayList(Arrays.asList("david@corda.net", "alice@corda.net", "bob@corda.net", "charlie@corda.net", "olivia@corda.net", "peter@corda.net"))


    @Before
    fun setup() {
        network = MockNetwork(
                MockNetworkParameters(cordappsForAllNodes = listOf(
                    TestCordapp.findCordapp("net.corda.samples.secretsanta.contracts"),
                    TestCordapp.findCordapp("net.corda.samples.secretsanta.flows")
                ),
                        networkParameters = testNetworkParameters,
                        notarySpecs = listOf(MockNetworkNotarySpec(CordaX500Name("Notary","London","GB")))

                ))
        santa = network.createNode(MockNodeParameters())
        elf = network.createNode(MockNodeParameters())
        val startedNodes = arrayListOf(santa, elf)
        network.runNetwork()
    }

    @After
    fun tearDown() {
        network.stopNodes()
    }

    @Test
    @Throws(Exception::class)
    fun flowUsesCorrectNotary() {
        val f1 = CreateSantaSessionFlow(playerNames, playerEmails, elf!!.info.legalIdentities[0])
        val future = santa!!.startFlow<SignedTransaction>(f1)
        network!!.runNetwork()
        val signedTransaction = future.get()
        assertEquals(1, signedTransaction.tx.outputStates.size)
        // ensure correct notary is used
        assertEquals(network!!.notaryNodes[0].info.legalIdentities[0], signedTransaction.notary)
    }

    @Test
    @Throws(Exception::class)
    fun canCreateSession() {
        val f1 = CreateSantaSessionFlow(playerNames, playerEmails, elf!!.info.legalIdentities[0])
        val future = santa!!.startFlow<SignedTransaction>(f1)
        network!!.runNetwork()
        val signedTransaction = future.get()
        assertEquals(1, signedTransaction.tx.outputStates.size)
        val output = signedTransaction.tx.outputsOfType(SantaSessionState::class.java)[0]
        // get some random data from the output to verify
        assertEquals(playerNames, output.playerNames)
    }

    @Test
    @Throws(Exception::class)
    fun transactionConstructedHasCorrectOutput() {
        val f1 = CreateSantaSessionFlow(playerNames, playerEmails, elf!!.info.legalIdentities[0])
        val future = santa!!.startFlow<SignedTransaction>(f1)
        network!!.runNetwork()
        val signedTransaction = future.get()
        assertEquals(1, signedTransaction.tx.outputStates.size)
        val (_, _, notary) = signedTransaction.tx.outputs[0]
        // ensure correct notary is used is used
        assertEquals(network!!.notaryNodes[0].info.legalIdentities[0], notary)
        val output = signedTransaction.tx.outputsOfType(SantaSessionState::class.java)[0]
        // checking player names, emails, and assignments.
        assertEquals(playerNames, output.playerNames)
        assertEquals(playerEmails, output.playerEmails)
        assertEquals(playerEmails.size, output.getAssignments()!!.size)
    }

    @Test
    @Throws(Exception::class)
    fun transactionConstructedHasOneOutputUsingTheCorrectContract() {
        val f1 = CreateSantaSessionFlow(playerNames, playerEmails, elf!!.info.legalIdentities[0])
        val future = santa!!.startFlow<SignedTransaction>(f1)
        network!!.runNetwork()
        val signedTransaction = future.get()
        assertEquals(1, signedTransaction.tx.outputStates.size)
        val (_, contract) = signedTransaction.tx.outputs[0]
        assertEquals("net.corda.samples.secretsanta.contracts.SantaSessionContract", contract)
    }

    @Test
    @Throws(Exception::class)
    fun transactionConstructedByFlowHasOneIssueCommand() {
        val f1 = CreateSantaSessionFlow(playerNames, playerEmails, elf!!.info.legalIdentities[0])
        val future = santa!!.startFlow<SignedTransaction>(f1)
        network!!.runNetwork()
        val signedTransaction = future.get()
        assertEquals(1, signedTransaction.tx.commands.size)
        val (value) = signedTransaction.tx.commands[0]
        assert(value is SantaSessionContract.Commands.Issue)
    }

    @Test
    @Throws(Exception::class)
    fun transactionConstructedByFlowHasOneCommandWithTheIssuerAndTheOwnerAsASigners() {
        val f1 = CreateSantaSessionFlow(playerNames, playerEmails, elf!!.info.legalIdentities[0])
        val future = santa!!.startFlow<SignedTransaction>(f1)
        network!!.runNetwork()
        val signedTransaction = future.get()
        assertEquals(1, signedTransaction.tx.commands.size)
        val (_, signers) = signedTransaction.tx.commands[0]
        assertEquals(2, signers.size)
        assertTrue(signers.contains(santa!!.info.legalIdentities[0].owningKey))
        assertTrue(signers.contains(elf!!.info.legalIdentities[0].owningKey))
    }

    @Test
    @Throws(Exception::class)
    fun transactionConstructedByFlowHasNoInputsAttachmentsOrTimeWindows() {
        val f1 = CreateSantaSessionFlow(playerNames, playerEmails, elf!!.info.legalIdentities[0])
        val future = santa!!.startFlow<SignedTransaction>(f1)
        network!!.runNetwork()
        val signedTransaction = future.get()
        assertEquals(0, signedTransaction.tx.inputs.size)
        assertEquals(1, signedTransaction.tx.outputs.size)
        // The single attachment is the contract attachment.
        assertEquals(1, signedTransaction.tx.attachments.size)
        Assertions.assertNull(signedTransaction.tx.timeWindow)
    }
}
